soccer/gameplay/skills/capture.py

```diff
@@ -1,270 +1,302 @@
+        # On the unit circle, two vectors are within 90 degrees of each other if the distance
+        # between two vectors is less than sqrt(1^2 + 1^2)
+        #
+        #              +1 | a
+        #                 |
+        #                 |
+        #                 |            b
+        #                 |_____________
+        #                              +1
+        #
+        # Dist between a and b are sqrt(2)
+        #
+        #
+        #     a                         b
+        #     ____________________________
+        #     -1                        +1
+        #
+        # Dist between a and b is 2
+        return (vec1.norm() - vec2.norm()).mag() < math.sqrt(2)
+
+    def on_enter_captured(self):
+        print("Enter captured")
+
+    def execute_captured(self):
+        self.update_held_cnt()
+        self.frames_in_captured += 1
+
+    def on_enter_settle(self):
+        print("Enter settle")
+        self.remove_all_subbehaviors()
+        self.add_subbehavior(skills.settle.Settle(),
+                             name='settler',
+                             required=True,
+                             priority=10)
+
+    def on_enter_collect(self):
+        print("Enter collect")
+        self.remove_all_subbehaviors()
+        self.add_subbehavior(skills.collect.Collect(),
+                             name='collector',
+                             required=True,
+                             priority=10)
+
+    def on_enter_completed(self):
+        self.probably_held_cnt = Capture.PROBABLY_HELD_START
+        self.remove_all_subbehaviors()
+
+
+    def execute_completed(self):
+        self.update_held_cnt()
 
-    def execute_fine_approach(self):
         self.robot.disable_avoid_ball()
-        self.robot.set_dribble_speed(self.dribbler_power)
+        self.robot.set_dribble_speed(Capture.DRIBBLE_SPEED)
 
-        bot2ball_dir = (main.ball().pos - self.robot.pos).normalized()
+    def update_held_cnt(self):
+        if (self.robot is None):
+            return
 
-        approach = self.bot_to_ball() * Capture.FineApproachDistanceMultiplier + \
-                    bot2ball_dir * Capture.FineApproachMinDeltaSpeed + \
-                    main.ball().vel * Capture.FineApproachBallSpeedMultiplier
-        if (approach.mag() > 1):
-            approach = approach.normalized() * 1
-        self.robot.set_world_vel(approach)
+        # If we hold the ball, increment up to max
+        # if not, drop to 0
+        if (evaluation.ball.robot_has_ball(self.robot)): #self.robot.has_ball()):
+        #if (self.robot is not None and self.robot.vel.mag() < Collect.STOP_SPEED):
+            self.probably_held_cnt = min(self.probably_held_cnt + 1, Capture.PROBABLY_HELD_HISTORY_LENGTH)
+        else:
+            self.probably_held_cnt = max(self.probably_held_cnt - 1, 0)
 
-    def on_enter_delay(self):
-        self.start_time = time.time()
+    @staticmethod
+    def role_assignment_cost(robot):
```

Comment:
Check this math out too. My goal was to halfway estimate which robot will reach the ball first and choose that one